### PR TITLE
Ready to Go: Pledge Summary

### PR DESCRIPTION
**Pull request by lomky**
Ready to go now!

Finishes #709.

----

Previous bug fixing:
I am trying to finalize the functionality for #709 

The code is syntactically correct, but in testing I always get back {pledged: 0, sent: 0} even after setting up `Patients` with upcoming appointments, soft pledges, and hard pledges.

My current guess is a scoping issue or an Active Record mistake.

Test output:
```
Sat Nov 05 11:31:21 [master]~/repos/personal/dcaf_case_management $ rake test --verbose test/models/patient_test.rb
Started with run options --verbose --seed 6310

 FAIL["test_0001_should return sent and remaining pledge for the next 7 days", #<Class:0x0055abf3a341d0>, 3.5770852970017586]
 test_0001_should return sent and remaining pledge for the next 7 days#PatientTest::pledge_summary (3.58s)
        --- expected
        +++ actual
        @@ -1 +1,2 @@
        -"{:pledged=>300, :sent=>0}"
        +# encoding: US-ASCII
        +"{:pledged=>0, :sent=>0}"
        test/models/patient_test.rb:103:in `block (2 levels) in <class:PatientTest>'

  36/36: [=========================================================================================================================] 100% Time: 00:00:03, Time: 00:00:03

Finished in 3.57897s
36 tests, 54 assertions, 1 failures, 0 errors, 0 skips
```

To run the tests, grab a copy of the current state of my repo clone.

It relates to the following issue #s: 
* Bumps #709

### CODE REVIEW

File: test/models/patient_test.rb [hunk 1]
```diff
@@ -75,9 +78,34 @@ class PatientTest < ActiveSupport::TestCase
       assert @patient.valid?
       @patient.appointment_date = '2016-07-01'
       assert @patient.valid?
+      @patient.appointment_date = ( Date.today + 4 )
```
Comment (colinxfleming) [inline test/models/patient_test.rb]: Ah, here's a gotcha -- so one of the best/worst/best features of rails' testing library (Minitest) is that it's super good about isolating parts of your tests and running them in isolation, and then resetting everything afterwards. (This is one of the more domain-specific-y patterns of minitest.) I don't know much about non-ruby testing libraries so please forgive me if I'm pointing out something shared all over the place! A common pattern that you'll see goes like this:

```
describe 'appointment date behavior or some other cool feature' do
  before do
    # some setup stuff shared between tests in the describe block
    @patient = create :patient
  end

  it 'should be able to save an appointment' do
    @patient.appointment_date = '2016-11-10'
    @patient.save
    assert_equal '2016-11-10', @patient.appointment_date
  end

  it 'should not have an appointment on create' do
    assert_nil @patient.appointment_date
  end
end
```

In this paradigm, the `before` block gets executed before every test runs, so we run some common setup that's available to every test within the `describe`. (What's more, Minitest runs tests in a random order, so even if they did run in sequence here they'd still pass because `@patient` gets reset every time. And you'll notice that up top, on line 4, there's a `before` block for the entire test class - this gets run before every single `describe` block as well.

So in a nutshell, these lines aren't making it to the `pledge_summary` describe block. Chuck 'em in a `before` in the pledge summary describe and you should be in business! e.g.:

```
describe 'pledge summary' do
  before do
    @patient.update appointment_date: (Date.today + 4)
    # ... and so on
  end
  
  it 'should return sent and remaining pledge...'
     .....
```

sorry for the super long winded explanation -- hope it was more helpful than confusing!

File: test/models/patient_test.rb [hunk 2]
```diff
@@ -75,9 +78,34 @@ class PatientTest < ActiveSupport::TestCase
...
   end
 
+  describe 'pledge_summary' do
+      it "should return sent and remaining pledge for the next 7 days" do
+          summary = Patient.pledged_status_summary
+          assert_equal '{:pledged=>0, :sent=>0}', summary.to_s
+          [@patient, @patient2].each do |pt|
+            create :pregnancy, patient: pt, created_by: @user
+          end
+          @patient.pregnancy.dcaf_soft_pledge = 300
```
Comment (colinxfleming) [inline test/models/patient_test.rb]: slightly quicker one-liner: `@patient.pregnancy.update attribute: value`. That'll update the value and save it in one fell swoop, so you don't have to `.save` on a separate line!

**Comment by lomky** [comment]
@colinxfleming That did it! Thank you very much for the review. It was completely understandable and helped me understand RoR better!

Assuming CircleCI passes, I think this is good to go.

**Comment by lomky** [comment]
I do not understand the tests failing :( It doesn't seem to be in code I changed, and it only happens on a full test run:

```
Sun Nov 06 13:10:57 [master]~/repos/personal/dcaf_case_management $ bundle exec "rake test"
Started with run options --seed 58363

581f723fae611223d193f47c,581f723fae611223d193f481,581f723fae611223d193f477,581f723fae611223d193f472                                  ] 2% Time: 00:00:00,  ETA: 00:00:38
 FAIL["test_0001_should let you save more than one call", #<Class:0x00559f1bf92670>, 65.60973407499841]
 test_0001_should let you save more than one call#LoggingCallsTest::logging multiple calls (65.61s)
        Expected false to be truthy.
        test/integration/logging_calls_test.rb:75:in `block (3 levels) in <class:LoggingCallsTest>'
        test/integration/logging_calls_test.rb:74:in `block (2 levels) in <class:LoggingCallsTest>'

click/193: [============================================================================================                            ] 77% Time: 00:01:25,  ETA: 00:00:25
  250/250: [=======================================================================================================================] 100% Time: 00:01:39, Time: 00:01:39

Finished in 99.18490s
250 tests, 509 assertions, 1 failures, 0 errors, 0 skips
{"error": {"reason": "Please provide the repository token to upload reports via `-t :repository-token`", "context": null}, "meta": {"status": 400}}

Sun Nov 06 13:12:48 [master]~/repos/personal/dcaf_case_management $ rake test test/integration/logging_calls_test.rb 
Started with run options --seed 13058

  8/8: [===========================================================================================================================] 100% Time: 00:00:26, Time: 00:00:26

Finished in 26.74662s
8 tests, 27 assertions, 0 failures, 0 errors, 0 skips
{"error": {"reason": "Please provide the repository token to upload reports via `-t :repository-token`", "context": null}, "meta": {"status": 400}}

Sun Nov 06 13:13:35 [master]~/repos/personal/dcaf_case_management $ bundle exec "rake test test/integration/logging_calls_test.rb"
Started with run options --seed 3509

  8/8: [===========================================================================================================================] 100% Time: 00:00:26, Time: 00:00:26

Finished in 26.53799s
8 tests, 27 assertions, 0 failures, 0 errors, 0 skips
{"error": {"reason": "Please provide the repository token to upload reports via `-t :repository-token`", "context": null}, "meta": {"status": 400}}
```

**Comment by colinxfleming** [comment]
That test (logging calls) and one other one (update patient info) is a
scrub that tends to pass locally but choke in CI. 99% chance it's not you,
don't stress.

File: app/models/patient.rb [hunk 3]
```diff
@@ -100,6 +100,21 @@ def self.trim_urgent_patients
     end
   end
 
+  def self.pledged_status_summary (num_days=7)
+    # return pledge totals for patients with appts in the next num_days
+    # TODO move to Pledge class, when implemented?
+    outstanding_pledges = 0
+    sent_total = 0
+    Patient.where(:appointment_date.lte => Date.today + num_days).each do |patient|
```
Comment (colinxfleming) [inline app/models/patient.rb]: Nice work on this -- good job getting the activerecord call right, especially!

File: test/models/patient_test.rb [hunk 4]
```diff
@@ -78,6 +81,31 @@ class PatientTest < ActiveSupport::TestCase
     end
   end
 
+  describe 'pledge_summary' do
+    it "should return sent and remaining pledge for the next 7 days" do
+      # Doesn't error when there are no pregnancies
```
Comment (colinxfleming) [inline test/models/patient_test.rb]: This is kind of a separate feature of this method, no? Might be worth chucking in its own `it` block -- something like:

```
it 'should not choke if there are no pregnancies that week' do
  # destroy all existing patients and their embedded pregnancies
  assert_equal ...
end
```
 maybe?

File: test/models/patient_test.rb [hunk 5]
```diff
@@ -78,6 +81,31 @@ class PatientTest < ActiveSupport::TestCase
...
+      summary = Patient.pledged_status_summary
+      assert_equal '{:pledged=>0, :sent=>0}', summary.to_s
+      [@patient, @patient2].each do |pt|
+        create :pregnancy, patient: pt, created_by: @user
+      end
+      @patient.update appointment_date: ( Date.today + 4 )
+      @patient.pregnancy.update dcaf_soft_pledge: 300
+      @patient2.update appointment_date: ( Date.today + 8 )
+      @patient2.pregnancy.update dcaf_soft_pledge: 500
```
Comment (colinxfleming) [inline test/models/patient_test.rb]: You can one-liner these two transformations to the pregnancy! `...pregnancy.update dcaf_soft_pledge: 500, pledge_sent: true` will do 'em both in one fell swoop. Update takes a hash and sets stuff with those keys to the values. You might be able to also one-liner the appointment date as `patient.update appointment_date: (Date.today + 4), pregnancy: { dcaf_soft_pledge: ... }` but I forget whether that's something that works or whether that's something that I think should work and doesn't 💯

File: test/models/patient_test.rb [hunk 6]
```diff
@@ -78,6 +81,31 @@ class PatientTest < ActiveSupport::TestCase
...
+      @patient2.pregnancy.update pledge_sent: true
+      # Returns 0 when there are no scope'd pregnancies
+      summary = Patient.pledged_status_summary(1)
+      assert_equal '{:pledged=>0, :sent=>0}', summary.to_s
```
Comment (colinxfleming) [inline test/models/patient_test.rb]: Rather than reassigning the summary variable, what do you think about just straight up comparing `assert_equal (the hash), Patient.pledged_status_summary(1)`?

File: test/models/patient_test.rb (hunk 4 above)
Comment (lomky) [inline test/models/patient_test.rb]: How would one `destroy all existing patients and their embedded pregnancies` within that test block?

**Comment by colinxfleming** [comment]
Since everything gets rebuilt from scratch in every `it` block, you can delete anything within a test and it'll be rebuilt for the next test. Try `Patient.destroy_all` at the top of this particular test -- that'll run a command equivalent to `delete from patients` (or whatever the mongodb equivalent is, I forget exactly).

ps let me know if this isn't helpful or if it's making it too overly complicated -- this is pretty much good to and we're well into optimizing territory! 💯

File: test/models/patient_test.rb (hunk 4 above)
Comment (colinxfleming) [inline test/models/patient_test.rb]: Since everything gets rebuilt from scratch in every it block, you can delete anything within a test and it'll be rebuilt for the next test. Try Patient.destroy_all at the top of this particular test -- that'll run a command equivalent to delete from patients (or whatever the mongodb equivalent is, I forget exactly).

ps let me know if this isn't helpful or if it's making it too overly complicated -- this is pretty much good to and we're well into optimizing territory! 💯

**Comment by lomky** [comment]
We're definitely into the academic realm, but I really appreciate it! I
don't get to develop with others very often anymore so the review
experience is refreshing. :)

I'll get you one last commit in the morning and then it will be :100:

**Comment by colinxfleming** [comment]
Good deal, glad it's still helpful! Very glad to have you on the team with us -- let me know if you have any other questions.

**Comment by lomky** [comment]
Good to Go! :fireworks:

**Comment by colinxfleming** [comment]
Awesome work @lomky ! Thanks again for seeing this over the finish line, and for going the extra mile and writing tests. 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯
